File: code/graphs/wgraph-adt.java
class Edge {
    int to; long w;
    Edge(int t, long weight) { to = t; w = weight; }
}

class WGraph {
    int V;                                  // number of vertices
    List<List<Edge>> adj;                   // adj.get(u) = u's out-edges
    WGraph(int n) {
        V = n;
        adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
    }
    void addEdge(int u, int v, long w) {
        adj.get(u).add(new Edge(v, w));
        // adj.get(v).add(new Edge(u, w));  // add this line for an UNDIRECTED edge
    }
}

File: code/graphs/wgraph-adt.cpp
struct Edge {
    int  to;    // the neighbor this edge leads to
    long w;     // the edge's weight (the cost of traversing it)
};

struct WGraph {
    int V;                            // number of vertices
    vector<vector<Edge>> adj;         // adj[u] = u's weighted out-edges
    WGraph(int n) : V(n), adj(n) {}
    void addEdge(int u, int v, long w) {
        adj[u].push_back({v, w});
        // adj[v].push_back({u, w});  // add this line for an UNDIRECTED edge
    }
};
